From cc098dbe94926a3ff8a4a2c58dab0da9e611d5dd Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Wed, 20 Dec 2023 14:42:16 -0800
Subject: [PATCH] wifi: mac80211: Add rtnl locks before scheduling
 autodisconnect

The cfg80211_autodisconnect worker threads have the rtnl lock and then
do a stop_ap bringing down the device and then release the rtnl lock,
which fails to destroy the bonded device due to deadlocks.
Add rtnl locks before scheduling the cfg80211_autodisconnect work queue
as this will ensure all the worker threads to disconnect with the lock
held and then release after all the slaves are destroyed.

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 net/wireless/nl80211.c | 2 ++
 net/wireless/sme.c     | 2 --
 2 files changed, 2 insertions(+), 2 deletions(-)

--- a/net/wireless/sme.c
+++ b/net/wireless/sme.c
@@ -22,6 +22,7 @@
 #include "reg.h"
 #include "rdev-ops.h"
 
+extern int g_bonded_interface_model;
 /*
  * Software SME in cfg80211, using auth/assoc/deauth calls to the
  * driver. This is for implementing nl80211's connect/disconnect
@@ -1584,7 +1585,8 @@ void cfg80211_autodisconnect_wk(struct w
 		container_of(work, struct wireless_dev, disconnect_wk);
 	struct cfg80211_registered_device *rdev = wiphy_to_rdev(wdev->wiphy);
 
-	rtnl_lock();
+	if (g_bonded_interface_model)
+		rtnl_lock();
 	wiphy_lock(&rdev->wiphy);
 	wdev_lock(wdev);
 
@@ -1625,5 +1627,6 @@ void cfg80211_autodisconnect_wk(struct w
 
 	wdev_unlock(wdev);
 	wiphy_unlock(&rdev->wiphy);
-	rtnl_unlock();
+	if (g_bonded_interface_model)
+		rtnl_unlock();
 }
